File: src/main/java/org/c3lang/intellij/stubs/C3ConstDeclarationStmtElementType.java
package org.c3lang.intellij.stubs;

import com.intellij.psi.PsiElement;
import com.intellij.psi.stubs.IndexSink;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import org.c3lang.intellij.index.NameIndex;
import org.c3lang.intellij.psi.C3ConstDeclarationStmt;
import org.c3lang.intellij.psi.C3StubElementType;
import org.c3lang.intellij.psi.C3StubElementTypeFactory;
import org.c3lang.intellij.psi.impl.C3ConstDeclarationStmtImpl;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class C3ConstDeclarationStmtElementType extends C3StubElementType<C3ConstDeclarationStmtStub, C3ConstDeclarationStmt>
{
	private static final C3ConstDeclarationStmtElementType INSTANCE = new C3ConstDeclarationStmtElementType();

	private C3ConstDeclarationStmtElementType()
	{
		super(C3StubElementTypeFactory.CONST_DECLARATION_STMT);
	}

	public static C3ConstDeclarationStmtElementType getInstance()
	{
		return INSTANCE;
	}

	@Override
	public @NotNull C3ConstDeclarationStmt createPsi(@NotNull C3ConstDeclarationStmtStub stub)
	{
		return new C3ConstDeclarationStmtImpl(stub, this);
	}

	@Override
	public @NotNull C3ConstDeclarationStmtStub createStub(@NotNull C3ConstDeclarationStmt psi, @NotNull StubElement<? extends PsiElement> stubElement)
	{
		return new C3ConstDeclarationStmtStub(stubElement, this, psi);
	}

	@Override
	public void serialize(@NotNull C3ConstDeclarationStmtStub stub, @NotNull StubOutputStream dataStream) throws IOException
	{
		stub.serialize(dataStream);
	}

	@Override
	public @NotNull C3ConstDeclarationStmtStub deserialize(@NotNull StubInputStream dataStream, StubElement stubElement) throws IOException
	{
		return new C3ConstDeclarationStmtStub(stubElement, this, dataStream);
	}

	@Override
	public void indexStub(@NotNull C3ConstDeclarationStmtStub stub, @NotNull IndexSink sink)
	{
		sink.occurrence(NameIndex.KEY, stub.getName().getFullName());
	}
}

File: src/main/java/org/c3lang/intellij/stubs/C3ConstDeclarationStmtElementType.kt
package org.c3lang.intellij.stubs

import com.intellij.psi.PsiElement
import com.intellij.psi.stubs.IndexSink
import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import org.c3lang.intellij.index.NameIndex
import org.c3lang.intellij.psi.C3ConstDeclarationStmt
import org.c3lang.intellij.psi.C3StubElementType
import org.c3lang.intellij.psi.C3StubElementTypeFactory
import org.c3lang.intellij.psi.impl.C3ConstDeclarationStmtImpl
import java.io.IOException

class C3ConstDeclarationStmtElementType :
    C3StubElementType<C3ConstDeclarationStmtStub, C3ConstDeclarationStmt>(C3StubElementTypeFactory.CONST_DECLARATION_STMT) {

    override fun createPsi(stub: C3ConstDeclarationStmtStub): C3ConstDeclarationStmt {
        return C3ConstDeclarationStmtImpl(stub, this)
    }

    override fun createStub(psi: C3ConstDeclarationStmt, stubElement: StubElement<out PsiElement?>): C3ConstDeclarationStmtStub {
        return C3ConstDeclarationStmtStub(stubElement, this, psi)
    }

    @Throws(IOException::class)
    override fun serialize(stub: C3ConstDeclarationStmtStub, dataStream: StubOutputStream) {
        stub.serialize(dataStream)
    }

    override fun indexStub(stub: C3ConstDeclarationStmtStub, sink: IndexSink) {
        sink.occurrence(NameIndex.KEY, stub.name.fullName)
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, stubElement: StubElement<*>): C3ConstDeclarationStmtStub {
        return C3ConstDeclarationStmtStub(
            stubElement,
            this,
            dataStream
        )
    }

    companion object {
        @JvmStatic
        val instance: C3ConstDeclarationStmtElementType = C3ConstDeclarationStmtElementType()
    }
}